### PR TITLE
LDAP Injection detection

### PR DESCRIPTION
**Pull request by jandro996**
# What Does This Do

Add support for LDAP injection detection by instrumenting calls to javax.naming.directory.InitialDirContext: 

- [search](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/InitialDirContext.html#search(java.lang.String,%20java.lang.String,%20javax.naming.directory.SearchControls))([String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) name, [String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) filter, [SearchControls](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/SearchControls.html) cons)(url)
- [search](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/InitialDirContext.html#search(java.lang.String,%20java.lang.String,%20java.lang.Object[],%20javax.naming.directory.SearchControls))([String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) name, [String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) filterExpr, [Object](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html)[] filterArgs, [SearchControls](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/SearchControls.html) cons)
- [search](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/InitialDirContext.html#search(javax.naming.Name,%20java.lang.String,%20java.lang.Object[],%20javax.naming.directory.SearchControls))([Name](https://docs.oracle.com/javase/7/docs/api/javax/naming/Name.html) name, [String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) filterExpr, [Object](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html)[] filterArgs, [SearchControls](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/SearchControls.html) cons)
- [search](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/InitialDirContext.html#search(javax.naming.Name,%20java.lang.String,%20javax.naming.directory.SearchControls))([Name](https://docs.oracle.com/javase/7/docs/api/javax/naming/Name.html) name, [String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) filter, [SearchControls](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/SearchControls.html) cons)

# Additional Notes

This is a first versión, the way the evidence is shown can be improved for sinks with String filterExpr and Object[] filterArgs arguments. For example  for [search](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/InitialDirContext.html#search(javax.naming.Name,%20java.lang.String,%20java.lang.Object[],%20javax.naming.directory.SearchControls))([Name](https://docs.oracle.com/javase/7/docs/api/javax/naming/Name.html) name, [String](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) filterExpr, [Object](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html)[] filterArgs, [SearchControls](https://docs.oracle.com/javase/7/docs/api/javax/naming/directory/SearchControls.html) cons)

given:
name  = 'ou=system'   
filterExpr   = '(&(uid={0})(userPassword={1}))'
filterArgs = ['user', '**TAINTED**'] 

with this approach the evidence will be  'name (&(uid={0})(userPassword={1})) user **TAINTED**' this can be improved to generate an evidence like 'name (&(uid=user)(userPassword=**TAINTED**))'

### CODE REVIEW

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java [hunk 1]
```diff
@@ -285,13 +288,46 @@ public void onProcessBuilderStart(@Nullable final List<String> command) {
...
     }
     final TaintedObjects taintedObjects = ctx.getTaintedObjects();
-    checkCommandInjection(taintedObjects, command);
+    checkInjection(taintedObjects, command, VulnerabilityType.COMMAND_INJECTION);
+  }
+
+  @Override
+  public void onDirContextSearch(String name, @Nonnull String filterExpr, Object[] filterArgs) {
+    // TODO check if is worth to calc size and use String[] instead of List<String>
+    List<String> elements = new ArrayList<>();
```
Comment (smola) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java:297]: Can we avoid the temporary list here? As far as I see, we don't need to accumulate anything. We can just have a function to check an arg, and pass `name` first`, then `filterExpr`, then all `filterArgs`...?

File: settings.gradle [hunk 2]
```diff
@@ -197,6 +197,7 @@ include ':dd-java-agent:instrumentation:java-directbytebuffer'
 include ':dd-java-agent:instrumentation:java-lang'
 include ':dd-java-agent:instrumentation:java-lang:java-lang-9'
 include ':dd-java-agent:instrumentation:java-security'
+include ':dd-java-agent:instrumentation:javax-namig-directory'
```
Comment (devinsba) [inline settings.gradle:200]: I think you meant: 
```suggestion
include ':dd-java-agent:instrumentation:javax-naming-directory'
```

Comment (jandro996) [reply settings.gradle]: thanks! 👍

**Review by smola — APPROVED** [review]

File: settings.gradle [hunk 3]
```diff
@@ -202,6 +202,7 @@ include ':dd-java-agent:instrumentation:java-io'
 include ':dd-java-agent:instrumentation:java-lang'
 include ':dd-java-agent:instrumentation:java-lang:java-lang-9'
 include ':dd-java-agent:instrumentation:java-security'
+include ':dd-java-agent:instrumentation:javax-naming-directory'
```
Comment (manuel-alvarez-alvarez) [inline settings.gradle:205]: I'll rather only use the base package name: `javax-naming`

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java [hunk 4]
```diff
@@ -415,6 +419,49 @@ public void onPathTraversal(final @Nullable File parent, final @Nonnull String c
...
 
+  @Override
+  public void onDirContextSearch(String name, @Nonnull String filterExpr, Object[] filterArgs) {
+    List<String> elements = null;
+    if (canBeTaintedNullSafe(name)) {
+      elements = new ArrayList<>();
+      elements.add(name);
+    }
+    if (canBeTaintedNullSafe(filterExpr)) {
+      elements = getElements(elements);
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java:430]: A ternary operator makes it easier to read:

```java
elements = elements != null ? elements : new ArrayList<>();
```

At first glance I thought `getElements` was doing something more sophisticated

**Review by manuel-alvarez-alvarez — APPROVED** [review]

File: settings.gradle (hunk 3 above)
Comment (jandro996) [reply settings.gradle]: done

**Pull request merged**